text/trace/0226-sampling-random-traceids.md

```diff
@@ -0,0 +1,214 @@
+nearest base-2 floating point, which is exactly 0x1.999999999999ap-04,
+which is approximately 0.10000000000000000555.  The exact quantity in
+this example, 0x1.999999999999ap-04, is multipled by `0x1p+56` and
+rounded to an unsigned integer (7205759403792794).  This specification
+says that to carry out sampling probability "0.1", we should keep
+Traces whose least-significant 56 bits form an unsigned value less
+than 7205759403792794.
+
+## T-value encoding for adjusted counts
+
+The example used sampling probability "0.1", which is a concisely
+rounded value but not exactly a power of two.  The use of decimal
+floating point in this case conceals the fact that there is an integer
+reciprocal, and when there is an integer reciprocal there are good
+reasons to preserve it.  Rather than encoding "0.1", it is appealing
+to encode the adjusted count (i.e., "10") because it conveys exactly
+the user's intention.
+
+This suggests that the t-value encoding be designed to accept either
+the sampling probability or the adjusted count, depending on how the
+sampling probability was derived.  Thus, the proposed t-value shall be
+parsed as a floating point or integer number using any POSIX-supported
+printf format specifier.  Values in the range [0x1p-56, 0x1p+56] are
+valid.  Values in the range [0x1p-56, 1] are interpreted as a sampling
+probability, while values in the range [1, 0x1p+56] are intepreted as
+an adjusted count.  Adjusted count values must be integers, while
+sampling probability values can be arbitrary floating point values.
+
+Whether to encode sampling probabilty or adjusted count is a choice.
+In both cases, the interpreted value translates into an exact
+threshold, which determines the exact inclusion probability.  From the
+exact inclusion probability, we can determine the adjusted count to
+use in a span-to-metrics pipeline.  When the t-value is _stated_ as an
+adjusted count (as opposed to a sampling probabilty), implementations
+can use the integer value in a span-to-metrics pipeline.  Otherwise,
+implementations should use an adjusted count of 1 divided by the
+sampling probability.
+
+## Where to store t-value in a Span and/or Log Record
+
+As specified, t-value should be encoded in the TraceState field of the
+span.  Probabilistic head samplers should set t-value in propagated
+contexts so that children using ParentBased samplers are correctly
+counted.
+
+Although prepared as a solution for Head and Intermediate Span
+sampling, the t-value encoding scheme could also be used to convey
+Logs sampling.  This document proposes to add an optional `LogState`
+string to the OTLP LogRecord, defined identically to the W3C
+tracecontext `TraceState` field.
+
+## Re-sampling with t-value
+
+It is possible to re-sample spans that have already been sampled,
+according to their t-value.  This allows a processor to further reduce
+the volume of data it is sending by lowering the sampling threshold.
```

Comment:
I want to understand the expected user behavior here. Let's take this example:

- A head sampler is used with a sampling rate of say 10%, 
- Let's say an intermediate sampler is also specified with a sampling rate of say 5%.

Let's say 1000 spans are emitted. Let's say ~100 spans are sampled by the head sampler (based on the above 10% rate). Now when they come to the intermediate sampler, isn't the user expectation that only ~5 out of these ~100 spans are sampled (and the rest 95 discarded)? If my understanding is correct, I was thinking that the t-value would need to be updated for those non-discarded spans, however the below proposal seems to be based on a different user behavior.

Reply:
By my understanding, the head sampler with a rate of 10% will export 100 spans (of 1000), and then the 5% sampler will export 50 of those spans if it's a consistent sampler.

My latest update to this OTEP includes an "s-value" discussed in last week's SIG which would allow non-consistent independent sampling to be reported, so the 5% simple probability sampler could attach `s:20` or `s:0.05`.

Reply:
For the record, I agree with @kalyanaj that 5% sampling rate for intermediate or tail sampling should mean that the volume of spans after sampling is about 20 times smaller than the input volume. Otherwise, assuming possible different sampling rates for the spans from the input stream, the user cannot really predict what the effect of 5% sampling would be.